clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7801,6 +7801,29 @@ static bool checkSYCLAddIRAttributesMergeability(const AddIRAttrT &NewAttr,
   return false;
 }
 
+void Sema::CheckSYCLAddIRAttributesFunctionAttrConflicts(Decl *D) {
+  const auto *AddIRFuncAttr = D->getAttr<SYCLAddIRAttributesFunctionAttr>();
+  if (!AddIRFuncAttr || AddIRFuncAttr->args_size() == 0 ||
+      hasDependentExpr(AddIRFuncAttr->args_begin(), AddIRFuncAttr->args_size()))
+    return;
+
+  // If there are potentially conflicting attributes, we issue a warning.
+  if (const auto *ReqdWGSizeAttr = D->getAttr<ReqdWorkGroupSizeAttr>())
+    Diag(ReqdWGSizeAttr->getLoc(),
+         diag::warn_sycl_old_and_new_kernel_attributes)
+        << ReqdWGSizeAttr;
+
+  if (const auto *ReqdSGSizeAttr = D->getAttr<IntelReqdSubGroupSizeAttr>())
```

Comment:
Did I get it right that we issue a warning unconditionally if a kernel has any property and the three chosen attributes? Should we check which property is attached to a kernel?

Reply:
That is right. The warning is issued if `add_ir_attributes_function` has values, regardless of the values. https://github.com/intel/llvm/pull/6941#discussion_r987244612 goes into more detail, but in summary we do it to avoid needing the frontend knowing about select values in this attribute while the warning is there to alert the user that they are using analogous features.

Reply:
Ok, these three attributes are all that we can apply to SYCL kernel or is there more?

Reply:
So, when there is a new property which conflicts with existing attribute, we have to update clang each time?

Reply:
Sadly yes, though this should be rare after these, `device_has`, and any existing SYCL extensions with attributes, as we will be leaning towards adding compile-time properties extensions rather than extensions adding new attributes, avoiding any more conflicts.